# Problem Set 03: Python Basics II - comments/solutions

# Programming for Economists Autumn 2026: Class 3
**Matias BF Hall**

**Follows Lecture 03**

This problem set covers loops, functions, return values, optional arguments, scope, classes, attributes, and methods.

**Table of contents**<a id='toc0_'></a>    
1. [Trade Balances Across Countries](#toc1_)    
2. [A Government Budget Function](#toc2_)    
3. [An Economy as a Class](#toc3_)    
4. [Find and Fix the Errors in a Government Debt Path](#toc4_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=true
	minLevel=2
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

## 1. <a id='toc1_'></a>[Trade Balances Across Countries](#toc0_)

The lists below contain illustrative exports and imports in billions of DKK. The entries at the same position belong to the same country.

1. Create an empty list called `trade_balances`.
2. Use `zip(exports, imports)` in a `for` loop. Calculate exports minus imports for each country and append the result to `trade_balances`.
3. Use `enumerate(countries)` in a second loop to print the country, its position, and its trade balance.
4. Inside that second loop, count how many countries have a trade surplus. A trade surplus is a balance strictly above zero.
5. Print the complete list, the total trade balance, and the number of countries with a surplus.

In one sentence, explain why `zip` is useful when the two economic series must remain matched.

In [2]:
countries = ["Denmark", "Sweden", "Germany", "France"]
exports = [1350.0, 1220.0, 1800.0, 980.0]
imports = [1270.0, 1290.0, 1650.0, 1040.0]

In [7]:
# Solution with explanation

# 1.
trade_balance = [] # [] creates an empty list

# 2.
for ex, im in zip(exports, imports): # zip() combines the two lists into pairs of elements
    trade_balance.append(ex - im) # append() adds an element to the end of the list

# ALTERNATIVE SOLUTION:
# 1. AND 2. COMBINED:
trade_balance_list_comp = [ex - im for ex, im in zip(exports, imports)] #using list comprehension to create the list in one line. The code is one to one equivalent to the for loop above.
# this exercise is to get you familiar with the for loop, append() and zip() functions.
# zip() is really useful when you want to iterate over two or more lists at the same time, while making sure that you are using the same index for each list
# ALTERNATIVE SOLUTION 2:
# using numpy we can do this in one line without zip(), you will learn about numpy later, but here is the code for reference:
import numpy as np # usually we have imports in the beginning of the code, but here we import it just for reference
trade_balance_np = np.array(exports) - np.array(imports) # this creates two numpy arrays and subtracts them element-wise, returning a numpy array.
# Element wise operation is not possible with regular Python lists. You can convert it to a list with .tolist() if you want a list instead of a numpy array.
trade_balance_np_list = trade_balance_np.tolist() # this converts the numpy array to a list

# 3. and 4.
count = 0 # start counting from 0
for i, country in enumerate(countries): # enumerate() returns the index and the element of the list, i as the index and country as the element, are arbitrary names. Usually names that are descriptive of the data are used.
    if trade_balance[i] > 0: # if the trade balance is positive, add one to the count
        count += 1 
    print(f'{i:<5}{country}: {trade_balance[i]}') # '<5' formats the index to be left aligned with a width of 5 characters.

# 5.
print(f'Trade balance list: {trade_balance}') # print the trade balance list
print(f'Total trade balance: {sum(trade_balance)}') # print the total trade balance
print(f'Number of countries with a positive trade balance: {count}') # print the count of countries with positive trade balance

0    Denmark: 80.0
1    Sweden: -70.0
2    Germany: 150.0
3    France: -60.0
Trade balance list: [80.0, -70.0, 150.0, -60.0]
Total trade balance: 100.0
Number of countries with a positive trade balance: 2


**Why `zip` is useful:**

<details>
<summary><strong>Hints and Results</strong></summary>

- In the first loop, each round supplies one export value and its matching import value.
- Start the surplus counter at zero and increase it inside an `if` block.
- The country positions printed by `enumerate` are `0`, `1`, `2`, and `3`.
- `trade_balances` should be `[80.0, -70.0, 150.0, -60.0]`.
- The total trade balance is `100.0` billion DKK.
- Two countries have a trade surplus.

`zip` supplies the matching export and import entries together, reducing the risk of combining values from different countries.

</details>

## 2. <a id='toc2_'></a>[A Government Budget Function](#toc0_)

A government receives tax revenue and pays for government spending and interest on its debt. The *primary balance* excludes interest payments. The *overall balance* includes them.

Complete `budget_balances(...)` below.

1. Calculate the primary balance as tax revenue minus government spending.
2. Calculate the overall balance as the primary balance minus interest payments.
3. Return both balances.
4. Call the function without supplying `interest_payments`. Unpack and print both returned values.
5. Call it again using `interest_payments=80.0`. Unpack and print both values.
6. Change the global variable `interest_payments` below to `500.0` and repeat the first call without the optional argument. Explain why its result does not change.

The function must use only its arguments. Do not refer to variables outside the function body.

In [8]:
def budget_balances(tax_revenue, government_spending, interest_payments=0.0):
    '''Return the primary and overall government budget balances.'''

    # Complete the function here.
    primary_balance = tax_revenue - government_spending # task 1
    overall_balance = primary_balance - interest_payments # task 2

    return primary_balance, overall_balance # task 3

tax_revenue = 1120.0
government_spending = 1070.0
interest_payments = 80.0

# task 4
primary_no_interest, overall_no_interest = budget_balances(tax_revenue, government_spending) # unpack the returned tuple into two variables directly
print(f'With out supplying an interest rate(interest rate defaults to keyword argument = 0.0) the primary balance is: {primary_no_interest} and the overall balance is: {overall_no_interest}')

# task 5
with_interest = budget_balances(tax_revenue, government_spending, interest_payments = interest_payments) # store the returned tuple in a variable
print(f'With supplying an interest rate of {interest_payments} the primary balance is: {with_interest[0]} and the overall balance is: {with_interest[1]}') # unpack the tuple by indexing into it¨

# task 6
interest_payments = 500.0
no_interest_new = budget_balances(tax_revenue, government_spending) # store the returned tuple in a variable
print(f'After changing the variable "interest_payments" to {interest_payments}, the primary balance is: {no_interest_new[0]} and the overall balance is: {no_interest_new[1]}')
print('nothing changes because the function uses the default value of 0.0 for interest_payments, which is not affected by changing the variable outside the function.')
# nothing changes because the function uses the default value of 0.0 for interest_payments, which is not affected by changing the variable outside the function.

With out supplying an interest rate(interest rate defaults to keyword argument = 0.0) the primary balance is: 50.0 and the overall balance is: 50.0
With supplying an interest rate of 80.0 the primary balance is: 50.0 and the overall balance is: -30.0
After changing the variable "interest_payments" to 500.0, the primary balance is: 50.0 and the overall balance is: 50.0
nothing changes because the function uses the default value of 0.0 for interest_payments, which is not affected by changing the variable outside the function.


**Why the first call ignores the global variable:**  
Nothing changes because the function uses the default value of 0.0 for interest_payments, which is not affected by changing the variable outside the function.
The default value is a local varible that is created when the function is called. `interest_payments` outside the function is a global variable that is not used inside the function unless it is passed as an argument. When there are local and global variables with the same name, the local variable takes precedence inside the function.

<details>
<summary><strong>Hints and Results</strong></summary>

- Separate two returned values with a comma and unpack them into two variable names when calling the function.
- Without an interest argument, the default value is `0.0`. Both the primary and overall balances are therefore `50.0` billion DKK.
- With `interest_payments=80.0`, the primary balance remains `50.0`, while the overall balance is `-30.0` billion DKK.
- Changing the global variable to `500.0` does not affect a call that omits the optional argument. The function uses its local parameter, whose default is still `0.0`.

</details>

## 3. <a id='toc3_'></a>[An Economy as a Class](#toc0_)

The class below should keep several facts about an economy together and calculate two commonly used ratios. All figures are rounded and illustrative.

1. Complete `__init__` so that it stores the four inputs as attributes.
2. Complete `gdp_per_capita()` so that it returns GDP divided by population.
3. Complete `debt_to_gdp()` so that it returns government debt as a percentage of GDP.
4. Create an object called `denmark` using the values below.
5. Print its name, GDP per capita, and debt-to-GDP ratio.
6. Change the object's GDP attribute to `3_150_000_000_000` DKK and print both calculated values again.

In one sentence, explain what `self` refers to when the methods are called on `denmark`.

In [ ]:
class Economy:

    def __init__(self, name, gdp, population, government_debt):
        self.name = name
        self.gdp = gdp
        self.population = population
        self.government_debt = government_debt

    def __str__(self): # this is a built in method that is called when you print the object. It returns a string representation of the object.
        return f"Economy: {self.name}, GDP: {self.gdp:_}, Population: {self.population:_}, Government Debt: {self.government_debt:_}"
        # "_" formats the number to have "_" thousands separators.
    def gdp_per_capita(self):
        return self.gdp / self.population

    def debt_to_gdp(self):
        return self.government_debt / self.gdp


# Rounded and illustrative values
economy_name = "Denmark"
gdp = 3_000_000_000_000
population = 6_000_000
government_debt = 900_000_000_000

# Create the object and test its methods below.
denmark = Economy(economy_name, gdp, population, government_debt)
print('5.')
print(denmark)
print(f'Has a GDP per capita of: {denmark.gdp_per_capita():_.0f} and a debt to GDP ratio of: {denmark.debt_to_gdp():.2%}') # ".2%" formats the number as a percentage with 2 decimal places. ".0f" formats the number as a float with 0 decimal places.

# 6.
denmark.gdp = 3_150_000_000_000
print('6.')
print(denmark)
print(f'Has a GDP per capita of: {denmark.gdp_per_capita():_.0f} and a debt to GDP ratio of: {denmark.debt_to_gdp():.2%}') # ".2%" formats the number as a percentage with 2 decimal places.

5.
Economy: Denmark, GDP: 3_000_000_000_000, Population: 6_000_000, Government Debt: 900_000_000_000
Has a GDP per capita of: 500_000 and a debt to GDP ratio of: 30.00%
6.
Economy: Denmark, GDP: 3_150_000_000_000, Population: 6_000_000, Government Debt: 900_000_000_000
Has a GDP per capita of: 525_000 and a debt to GDP ratio of: 28.57%


**What `self` refers to:**  
Self refers to the object itself, which is `denmark` in this case. It allows the methods to access everything stored in the object, including its attributes and other methods.

<details>
<summary><strong>Hints and Results</strong></summary>

- Store an input on the object with syntax such as `self.name = name`.
- Each method can use the attributes stored on the object through `self`.
- With the original values, GDP per capita is `500000.0` DKK and the debt-to-GDP ratio is `30.0` percent.
- After GDP rises to `3_150_000_000_000` DKK, GDP per capita is `525000.0` DKK and the debt-to-GDP ratio is approximately `28.57` percent.
- When a method is called on `denmark`, `self` refers to that particular `Economy` object.

</details>

## 4. <a id='toc4_'></a>[Find and Fix the Errors in a Government Debt Path](#toc0_)

The function below should calculate government debt after each year's budget deficit. A positive deficit raises debt, while a negative deficit represents a surplus and lowers debt.

The code runs, but it returns `[100, 10]` instead of the full debt path `[100, 110, 125, 120]`. It contains three mistakes.

1. Find and fix all three mistakes.
2. Fix one mistake at a time and rerun the cell after each change.
3. Write one sentence explaining each correction.

In [4]:
def calculate_debt_path(initial_debt, annual_budget_deficits):
    '''Return the initial debt and the debt after each annual deficit.'''

    debt = initial_debt
    debt_path = [initial_debt]

    for deficit in annual_budget_deficits:
        debt = debt + deficit
        debt_path.append(debt)
    return debt_path


initial_debt = 100
annual_budget_deficits = [10, 15, -5]

print("Government debt path:", calculate_debt_path(initial_debt, annual_budget_deficits))

Government debt path: [100, 110, 125, 120]


**Errors corrected:**

1. ***Variable name***: The variable `debt` was calculated but never used. Instead, `deficit` was appended to the list. The correct variable to append is `debt`, which represents the updated debt after adding the deficit.
2. ***Reference***: `debt` is set to `initial_debt + deficit`, causing the debt path to always update from the initial debt rather than the most recent debt. Instead we can set `debt = debt + deficit` to have it update every iteration of the for loop.
3. ***Indentation***: `return debt_path` was indented too far, causing the function to return after the first iteration of the loop. It should be aligned with the `for` loop to ensure the entire loop executes before returning.

<details>
<summary><strong>Hints and Results</strong></summary>

- The function should return only after the loop has processed every year.
- Each new debt level must start from the previous debt level, not repeatedly from the initial value.
- The list should receive the updated debt level, not the annual deficit itself.
- The correct debt path is `[100, 110, 125, 120]`.
- The final debt is `120`, which is the initial debt plus all three annual deficits: `100 + 10 + 15 - 5`.

</details>